src/crimson/osd/watch.cc

```diff
@@ -300,19 +303,21 @@ void Notify::do_notify_timeout()
   if (complete) {
     return;
   }
-  // it might be that `this` is kept alive only because of the reference
-  // a watcher stores and which is being removed by `cancel_notify()`.
-  // to avoid use-after-free we bump up the ref counter with `guard_ptr`.
-  [[maybe_unused]] auto guard_ptr = shared_from_this();
-  for (auto& watcher : watchers) {
-    logger().debug("canceling watcher cookie={} gid={} use_count={}",
-      watcher->get_cookie(),
-      watcher->get_watcher_gid(),
-      watcher->use_count());
-    watcher->cancel_notify(ninfo.notify_id);
-  }
-  std::ignore = send_completion(std::move(watchers));
```

Comment:
I think we can just do:

```
  std::ignore = send_completion(std::move(watchers))).then([guard_ptr] {});
```

Reply:
I'm not sure whether there is any assumption about the send message ordering in this case, but it worths noting that cross-core calls cannot guarantee FIFO in the target core. With the current implementation, the `future` returned by `Connection::send()` should be chained correctly before the next message.

Also see https://github.com/ceph/ceph/pull/51147#discussion_r1206155226